Convert variable names to snake_case

GearsDatapacks opened the pull request:
Closes #2978 
Let me summarise what I've done:

- BadName errors are now no longer reported in the lexer. The lexer treats all previous bad names as valid.
- The analyser now is responsible for validating names. It checks the name of variables (`let`, `use` and `const`), functions, parameters, types and type constructors.
- Finding an error in a variable name still declares it, to allow any references to that variable to still type-check correctly.
- When the analyser finds an invalid name, it adds it to a `bad_names` array, just like `unused_imports`.
- I've had to add more location information to certain ast nodes to allow proper renaming of variables without overwriting other parts of the source code 

The main problem I see with this approach is that it would be easy to miss validation of names in certain places, but I can't think of an easy way to check all names in the analyser.
This implementation also doesn't rename any references to invalid variables, but #2956 may need to be closed before that's possible/easy to add.

GearsDatapacks commented:
I've decided to mark this ready for review as I would like feedback on it and it seems more likely to get reviewed this way

lpil commented on compiler-core/src/error.rs:
```diff
@@ -2927,10 +2928,79 @@ See: https://tour.gleam.run/advanced-features/use/");
+                },
+
+                TypeError::BadUpName { location, name } => {
+                    let label = "This is not a valid upname";
```
"upname" doesn't mean anything to a Gleam programmer, it's an internal name to the compiler and could change at any time. 

Let's merge these 3 new error variants into one `BadName` variant and have an enum to say what the context is:

```rust
pub enum BadNameKind {
    Type,
    TypeVariable,
    CustomTypeVariant,
    Variable,
    Function,
}
```

GearsDatapacks replied:
> "upname" doesn't mean anything to a Gleam programmer, it's an internal name to the compiler and could change at any time.

Right, I figured that but I stuck with "upname" since that was used in error messages previously

lpil commented on compiler-core/src/language_server/tests/snapshots/gleam_core__language_server__tests__action__rename_invalid_bit_array_pattern.snap:
```diff
@@ -0,0 +1,7 @@
+---
+source: compiler-core/src/language_server/tests/action.rs
+expression: "apply_first_code_action(code, 1)"
```
Please put the original src as the expression so we can tell what the snapshot is for. It is quite challenging to debug quickly if you can't see what the input is!

lpil commented on compiler-core/src/type_/expression.rs:
```diff
@@ -1315,6 +1349,17 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
+                    let errors = self.check_valid_pattern_names(pattern);
+                    self.errors.extend(errors);
+                }
+            }
```
We should not traverse the patterns an extra time for this, rather it should be done during the existing pattern analysis.

lpil reviewed (commented):
Lovely work!! Thank you very much. I've left a handful of notes inline

GearsDatapacks commented:
Thanks for all the feedback! I know it's quite a big PR, and I'm still pretty new to this codebase. I'll go through and implement these as soon as possible.

GearsDatapacks commented:
I've made some changes:
- I've removed `BadNames` from the capnp schema
- I've extracted duplicate name-checking code into methods
  - Since these checks occur in three different structs, I wrote a trait to generalise that. I'm not sure if that was the best way to go about it, but it's the one that I thought of.
- The test files should now contain the source code :)

GearsDatapacks commented:
Thinking about this more, it might be better to implement `NameChecker` as a struct, which just holds references to the `errors` and `bad_names` vectors.

lpil commented on compiler-core/src/type_/tests/errors.rs:
```diff
@@ -1130,6 +1130,174 @@ const foo = 2"
 }
 
+#[test]
+fn invalid_const_name() {
```
Beautiful tests!!!

lpil commented on compiler-core/src/type_/name.rs:
```diff
@@ -0,0 +1,81 @@
+use super::{BadNameKind, Error};
+use heck::{ToSnakeCase, ToUpperCamelCase};
+
+pub trait NameChecker {
```
We don't define new traits, please remove this.

Could you move all the logic for deciding whether names are valid elsewhere- it's not specific to types so it should go in the type_ module namespace.

lpil reviewed (commented):
Looking great!!

GearsDatapacks commented:
I've now moved the name checking code into function in `name.rs` inside the `analyse` module

lpil commented on compiler-core/src/analyse/name.rs:
```diff
@@ -0,0 +1,116 @@
+    (errors, bad_names)
+}
+
+fn check_bad_name(location: SrcSpan, name: &EcoString, kind: BadNameKind) -> Option<Error> {
```
This name implies that checks it is a valid "bad name", which I found confusing. Could we give it a clearer name please

lpil commented on compiler-core/src/analyse/name.rs:
```diff
@@ -0,0 +1,116 @@
+use ecow::EcoString;
```
None of the functions in this module check that the names are of the formats they are intended to check for. Using regexes here would be good.

lpil commented on compiler-core/src/analyse/name.rs:
```diff
@@ -0,0 +1,116 @@
+use super::{BadNameKind, Error};
+use heck::{ToSnakeCase, ToUpperCamelCase};
+
+type BadName = (SrcSpan, EcoString);
```
This name isn't right, it's not a bad name. It's a correction or such. Let's use a struct for this.

lpil commented on compiler-core/src/analyse/name.rs:
```diff
@@ -0,0 +1,116 @@
+            error,
+            (
+                location,
+                EcoString::from(format!("_{}", name.to_snake_case())),
```
Favour `eco_format!` rather than building a string and then converting it to an ecostring

lpil commented on compiler-core/src/analyse/name.rs:
```diff
@@ -0,0 +1,116 @@
+}
+
+pub fn check_valid_argument(names: &ArgNames) -> (Vec<Error>, Vec<BadName>) {
+    let mut errors = Vec::new();
+    let mut bad_names = Vec::new();
```
Pass in these as mutable references to avoid allocating these vectors and then immediately emptying them into another vector

lpil reviewed (commented):
Thank you!

GearsDatapacks commented:
I've now cleaned up everything that you suggested

lpil commented on CHANGELOG.md:
```diff
@@ -21,6 +21,21 @@
   being attempted.
   ([Ameen Radwan](https://github.com/Acepie))
 
+- LSP now suggests a code a action to rename variables, types and functions when they
+  don't match the gleam naming conventions:
```
```suggestion
- The language server now suggests a code a action to rename variables, types and
  functions when they don't match the Gleam naming requirements:
```

lpil commented on compiler-core/src/analyse/name.rs:
```diff
@@ -0,0 +1,139 @@
+) -> Option<(Error, NameCorrection)> {
+    // Some of the internally generated variables (such as `_capture` and `_use0`)
+    // start with underscores, so we allow underscores here.
+    let valid_name_regex = Regex::new("^_?[a-z][a-z0-9_]*$").expect("Regex is correct");
```
This should be a static value rather than recomputing the regex on every name. See other uses of regex in the compiler for how we do this

lpil commented on compiler-core/src/analyse/name.rs:
```diff
@@ -0,0 +1,139 @@
+    pub correction: EcoString,
+}
+
+pub fn check_valid_name(
```
These functions do more than just check the name, they also return a correction. Let's move that out of these and keep them doing one thing.

Let's also make the error case `Err`.

lpil commented on compiler-core/src/analyse.rs:
```diff
@@ -957,6 +1004,13 @@ impl<'a, A> ModuleAnalyzer<'a, A> {
+        {
+            self.errors.push(error);
+            self.name_corrections.push(correction);
+        }
```
This 
```rust
if let Some((error, correct)) = .... {
    self.errors.push(error)
    self.name_corrections.push(correction)
}
```
is duplicated several times.

Let's make a helper for it:

```rust
self.check_name_case(name, location, NameKind::Type)
```

lpil commented on compiler-core/src/type_/error.rs:
```diff
@@ -449,6 +463,46 @@ pub enum LiteralCollectionKind {
 }
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum BadNameKind {
```
```suggestion
pub enum Named {
```

lpil commented on compiler-core/src/type_/error.rs:
```diff
@@ -449,6 +463,46 @@ pub enum LiteralCollectionKind {
+}
+
+impl BadNameKind {
+    pub fn to_string(self) -> EcoString {
```
```suggestion
    pub fn as_str(&self) -> &'static str {
```

lpil commented on compiler-core/src/error.rs:
```diff
@@ -2949,9 +2950,39 @@ See: https://tour.gleam.run/advanced-features/use/");
+                TypeError::BadName { location, name, kind } => {
+                    let kind_str = kind.to_string();
+                    let label = format!("This is not a valid {kind_str} name");
+                    let text = if kind.is_discard() {
```
Use pattern matching instead of `if/else`

lpil reviewed (commented):
Nearly there! Thank you

GearsDatapacks commented:
All done now.
- I've made the various refactors you suggested
- The `check_valid_name` functions have now been reworked into two functions: `check_name_case` and `correct_name_case`
  - `check_name_case` returns an error if the given name is invalid
  - `correct_name_case` returns a `NameCorrection` for the given name

lpil reviewed (approved):
This is fantastic!!! Thank you!
